### ISSUE TITLE
Safe usage of localStorage

### ISSUE THREAD

**Issue by callmemagnus** [SonarSource/sonar-javascript#623]
I had a look on the [JS rules filtered on localStorage](https://sonarcloud.io/organizations/default/rules#languages=js|q=localStorage) and didn't find anything like what I propose below.

DOM storages may be broken at different levels.

The existing cases I know of:
- `sessionStorage` and `localStorage` are null (chrome and firefox when DOM storage is disabled by configuration)
- `localStorage` is erased after the session (chrome specific configuration)
- `localStorage.setItem` may throw an error (Safari in private mode or quota exceeded)
- `localStorage` is "disabled" and invoking it throws an "Access Denied" error

Directly using `localStorage` is code smell. Sonar should help by highlight this problem as this may have unexpected consequences. 

When the usage of `localStorage` is not in a try/catch statement and it throws an Error, the rest of the JS file is ignored which may break the site completely although only cross-session DOM storage is broken.

My propositions for detection rules :

- all `localStorage` mentions should be fenced in a try/catch
- all `sessionStorage` usages should be fenced in a try/catch

Exemple bad code:

```
/*
 * the following throws "Access denied" on IE under specific configuration
 */
var storage = localStorage; 

/*
 * the following throws:
 * - "Quota exceeded"
 * - trying to call setItem on null (disabled DOM storage)
 * - Error 22 ? on safari in private mode (by memory)
 */
storage.setItem('key', 'value'); // throws
```

Correct implementation:
```
try {
  var storage = localStorage;
  storage.setItem(...)
} catch (e) {
  // handle the different error if required
}
```

More information on the IE specific issue:

Usually in corporate environment, Windows may be "badly" configured and break `localStorage`, see https://msdn.microsoft.com/en-us/library/bb250462(v=vs.85).aspx.

What do you think ?

**Comment by callmemagnus** [comment]
And of course this is also valid for the typescript sonar project.

**Comment by vilchik-elena** [comment]
Hi @callmemagnus,

Thanks for the good idea. In fact we have a rule https://sonarcloud.io/organizations/default/rules#rule_key=javascript%3AS3271 for that topic in SonarJS, but its implementation is quite limited. We gonna consider its extension or may be creation of a new rule.

**Comment by vilchik-elena** [comment]
Relates to S[ONARJS-512](https://jira.sonarsource.com/browse/SONARJS-512)

**Comment by callmemagnus** [comment]
hello @vilchik-elena,

I saw that one but I think it just brings attention of the devs to the fact that `localStorage` is used and not that it is safely used (at js execution time).

That rule serves the purpose of data storage's content over the proper usage.

**Comment by callmemagnus** [comment]
By the way, my proposition also covers the setter case:
```js
localStorage.key = value;
```

**Comment by vilchik-elena** [comment]
@callmemagnus We've created the ticket #711 to extend existing rule S3271. As soon as it's fixed we will see how we better apply your suggestion. Meantime I'm closing this issue.
